### DISCUSSION TITLE
tracking only specific class

### DISCUSSION THREAD

**Discussion by jbk3407** [ultralytics/ultralytics#10074] [Q&A]
Hi, I'm trying to track specific class ONLY(ex. 'person', 'phone')using my laptop's webcam.

From the below code I wrote, I'm having hard time figuring out how to use `annotated_frame = results[0].plot()`  for tracking only the chosen class.

I would like to get a sample code if possible. Thanks in advance :)

```
import cv2
from collections import Counter
from ultralytics import YOLO
import os
os.environ['KMP_DUPLICATE_LIB_OK']='True' # To solve OMP: Error #15


model = YOLO('yolov8n.pt')

# choose source [src=webcam /src2=video]
src = 0
src2 = './test_tracking_tennis.mp4'

cap = cv2.VideoCapture(src)

# Loop through the video frames
while cap.isOpened():
    # Read a frame from the video
    success, frame = cap.read()

    if success:
        # Run YOLOv8 inference on the frame
        results = model.track(frame, persist=True)

        # Visualize the results on the frame
        annotated_frame = results[0].plot()

        # Display the annotated frame
        cv2.imshow("YOLOv8 Tracking", annotated_frame)

        # Break the loop if 'q' is pressed
        if cv2.waitKey(1) & 0xFF == ord("q"):
            break
    else:
        # Break the loop if the end of the video is reached
        break

# Release the video capture object and close the display window
cap.release()
cv2.destroyAllWindows()

```

**Comment by pderrenger** [comment]
@jbk3407 hi! 😄 It looks like you're on a great track with using your laptop's webcam for object tracking. If you're interested in tracking specific classes only, such as 'person' or 'phone', you can filter the `Results` object based on the classes you're interested in before calling the `.plot()` method. You could use the `classes` parameter in the `track()` method to filter for specific classes based on their IDs. For classes like 'person' and 'phone', you'll need to know their class indices in your model (for COCO pretrained models, 'person'=0, 'phone' usually isn't directly labeled but you might look for 'cell phone'=67).

Here is a modified version of your loop that filters for the 'person' class only:

```python
while cap.isOpened():
    success, frame = cap.read()
    if success:
        # Track only 'person' class (class ID for 'person' is 0)
        results = model.track(frame, persist=True, classes=[0])
        
        # You can now generate and display the annotated frame as before
        annotated_frame = results[0].plot()

        cv2.imshow("YOLOv8 Tracking", annotated_frame)

        if cv2.waitKey(1) & 0xFF == ord("q"):
            break
    else:
        break

cap.release()
cv2.destroyAllWindows()
```

This code snippet specifies the `classes` argument in the `track()` method call to only track objects of class 'person'. Replace `[0]` with the relevant class indices you are interested in. Also, do note that the class ID I used (`0` for 'person') is based on the standard COCO dataset class IDs; please adjust according to your model if different. Hope that helps and happy tracking! 🚀

**Reply by jbk3407** [reply]
Thank you for your fast reply!! Using `results = model.track(frame, persist=True, classes=[0])` works very well. :) 
However, I'm trying to get multiple selected classes('person and 'cell phone' and maybe few more). Is there a way to use kwargs to do that?
Maybe something like:

`results = model.track(frame, persist=True, classes=[0] or [67])` <<<but this doesn't seems to work :(
`results = model.track(frame, persist=True, classes=[0], classes=[67])`<<<but this doesn't seems to work :(
could you help me to use kwargs to track person and cell phone and maybe few more?

Also, if I use  this `results = model.track(frame, persist=True, classes=[0])` kwargs option, is this tracking everything but just showing the bounding box of a 'person' class? or is it actually tracking just the person?

Thanks in advance :)

**Reply by pderrenger** [reply]
Hey! 😊 I'm glad to hear the tracking for specific classes is working out for you. To track multiple classes like 'person' and 'cell phone', you just need to include their class IDs in a single list for the `classes` argument. Here's how:

```python
results = model.track(frame, persist=True, classes=[0, 67])  # For 'person' and 'cell phone'
```

This way, it tracks and shows bounding boxes for both 'person' and 'cell phone'. Adding more class IDs to the list will include those classes in the tracking as well.

Regarding your second question, using `classes=[0]` actually instructs the model to track only 'person' objects. It doesn't track everything and then filter; it specifically focuses on the classes you've mentioned. So, it's indeed tracking just the person in this case. 

Hope this clears things up for you! Let me know if you have any more questions. 😄

**Comment by Zhaok-123** [comment]
hello,When I use the tracking algorithm, how can I only track a specific id, and other IDs only predict normally without using the tracking algorithm?

**Reply by glenn-jocher** [reply]
@PETE-wasd to track only specific classes or IDs while others are predicted normally, you can filter `results` by the desired class or ID before running tracking logic. For instance, modify the detected objects within `results[0].boxes` to include only the classes/IDs of interest. Let me know if you need further clarification or assistance with code!

**Comment by Zhaok-123** [comment]
Thank you for your reply. My main code is as follows. How can I specifically track only a specific id while other ids predict normally? Or is there any way for me to set tracking parameters for each ID separately?

```
    while cap.isOpened():
        success, frame = cap.read()
        if success:
            results = model.track(source=frame,
                                  persist=True,
                                  tracker="botsort.yaml",
                                  classes=[0],
                                  conf=conf,
                                  iou=iou
                                  )

            annotated_frame = results[0].plot()
            cv2.imshow("YOLO11 Tracking", annotated_frame)
            if cv2.waitKey(1) & 0xFF == ord("q"):
                break

        else:
            break
```

**Reply by glenn-jocher** [reply]
@PETE-wasd to specifically track only certain object IDs (e.g., 'person', 'phone'), you can filter the `results` object to apply custom processing for those IDs before plotting. Currently, it is not possible to set tracker parameters individually per ID, as tracking parameters are globally applied across all tracked objects. Let me know if you need further clarification or assistance!
